apiserver/facades/client/keymanager/keymanager.go

```diff
@@ -61,338 +65,448 @@ func (api *KeyManagerAPI) checkCanWrite(ctx context.Context, sshUser string) err
+	// eventually Juju could support adding keys for individual users. We have
+	// now partially wired this support up in the service and when adding keys
+	// we speicfy the user to which the keys should be added for.
+	//
+	// Currently the Juju client doesn't support adding keys for a specific user
+	// but for the user arg in params the client always sets this to "admin".
+	// We can't rely on this value at the moment and have to use the
+	// authenticated entity.
+
+	if err := api.checkCanWrite(ctx); err != nil {
 		return params.ErrorResults{}, apiservererrors.ServerError(err)
 	}
 	if err := api.check.ChangeAllowed(ctx); err != nil {
-		return params.ErrorResults{}, errors.Trace(err)
+		return params.ErrorResults{}, err
 	}
 	if len(arg.Keys) == 0 {
 		return params.ErrorResults{}, nil
 	}
 
-	// For now, authorised keys are global, common to all users.
-	sshKeys, currentFingerprints, err := api.currentKeyDataForAdd(ctx)
-	if err != nil {
-		return params.ErrorResults{}, apiservererrors.ServerError(fmt.Errorf("reading current key data: %v", err))
+	if api.authedUser.Kind() != names.UserTagKind {
+		return params.ErrorResults{
+			Results: []params.ErrorResult{
+				{
+					Error: apiservererrors.ParamsError(
+						params.CodeTagKindNotSupported,
+						"authorised user %q unsupported, can only accept users of kind %q",
+						api.authedUser, names.UserTagKind,
+					),
+				},
+			},
+		}, nil
 	}
 
-	// Ensure we are not going to add invalid or duplicate keys.
-	results := transform.Slice(arg.Keys, func(key string) params.ErrorResult {
-		fingerprint, comment, err := ssh.KeyFingerprint(key)
-		if err != nil {
-			return params.ErrorResult{Error: apiservererrors.ServerError(fmt.Errorf("invalid ssh key: %s", key))}
-		}
-		if jujuKeyCommentIdentifiers.Contains(comment) {
-			return params.ErrorResult{Error: apiservererrors.ServerError(fmt.Errorf("may not add key with comment %s: %s", comment, key))}
-		}
-		if currentFingerprints.Contains(fingerprint) {
-			return params.ErrorResult{Error: apiservererrors.ServerError(fmt.Errorf("duplicate ssh key: %s", key))}
-		}
-		currentFingerprints.Add(fingerprint)
-		sshKeys = append(sshKeys, key)
-		return params.ErrorResult{}
-	})
-
-	err = api.writeSSHKeys(sshKeys)
-	if err != nil {
-		return params.ErrorResults{}, apiservererrors.ServerError(err)
+	user, err := api.userService.GetUserByName(ctx, api.authedUser.Id())
+	switch {
+	case errors.Is(err, accesserrors.UserNameNotValid):
+		return params.ErrorResults{
+			Results: []params.ErrorResult{
+				{
+					Error: apiservererrors.ParamsError(
+						params.CodeUserInvalidName,
+						"invalid user name: %q",
+						arg.User,
+					),
+				},
+			},
+		}, nil
+	case errors.Is(err, accesserrors.UserNotFound):
+		return params.ErrorResults{
+			Results: []params.ErrorResult{
+				{
+					Error: apiservererrors.ParamsError(
+						params.CodeUserNotFound,
+						"user %q does not exist",
+						arg.User,
+					),
```

Comment:
This one is fine because you can only be admin to write. If that changes, this is dangerous.